quill-core/src/main/scala/io/getquill/quotation/Parsing.scala

```diff
@@ -698,6 +698,8 @@ trait Parsing {
       Delete(astParser(query))
     case q"$action.returning[$r](($alias) => $body)" =>
       Returning(astParser(action), identParser(alias), astParser(body))
+    case q"$action.returningRecord" =>
```

Comment:
What do you mean with the "overloading" option? If I understand the suggested snipped correctly, it checks if the lambda passed to `returning` is of type `T => T`. With that API users aren't able specify the return type.

What about introducing `.returningAs[R]` and don't touch parsing of `. returning` at all? Introducing an additional API seems to be less complex at implementation and more flexible for the user.

Reply:
@rolandjohann So I think `.returning(e => e)` should work, otherwise we'll get WTFs from users. It's an expectations thing. It's totally reasonable for a user to ask "So wait, why is this API for whole records, and this one for single fields? That's a totally arbitrary distinction."

Also, are you saying that overloading this method now forces **all** calls to `.returning` then to specify the generic parameter including the regular use-case `.returning(a -> a.id)`? That definitely calls for introducing `returningAs`! Is this the case?

Why can't we do something like this (see the bottom for the addition)?

```scala
case q"$action.returning[$r](($alias) => $body)" =>
    val tpe =
    body match {
      case Ident(TermName(n)) if (n == alias.name.toString) => Some(x.tpe)
      case _ => None
    }
  val argTpe =
    alias match {
      case q"$mods val $pat: $tpt = $expr" =>
        Some(tpt.tpe)
      case _ => 
        None
    }

  (tpe, argTpe) match {
    case (Some(a), Some(b)) if (a =:= b) =>
      ReturningRecord(astParser(action))
    case _ =>
      Returning(astParser(action), identParser(alias), astParser(body))
  }          
}
// Option A (if returning can be overloaded without verbosity increase)
//case q"$action.returning[$r](...$paramss)" if (paramss.length == 0) => 

// Option B (if returningAs needs to be introduced)
case q"$action.returningAs[$r](...$paramss)" =>
  ReturningRecord(astParser(action))
```

If there's something I missed here or if this is not as simple as I think, let me know.

Reply:
Sorry, just recognized that I missed a point you mentioned earlier.

We can go with overloading, for this we must introduce the additional signature at `QueryDsl`. Just using the old one and omit the lambda which normally defines the column to return will return a function type:
```scala
val q1: (Person => Person) => ctx.ActionReturning[Person, Person] = query[Person].insert(lift(Person(0L, "Roland", 29))).returning[Person]
val q2: ctx.ActionReturning[Person, Long] = query[Person].insert(lift(Person(0L, "Roland", 29))).returning(_.id)
```

Reply:
Yeah. That's what I was thinking.

Reply:
And we must define the type when calling to get Entity, too.

```scala
.returning[Person]
.returning(_.id)
```